### PR TITLE
ref(hc): Replica ApiToken into region silo

### PR DESCRIPTION
**Pull request by corps**
Replicates the ApiToken model into the region silo in support of https://github.com/getsentry/sentry/pull/57184

Notably, some complexity arises from the need to replicate ApiToken's organization_id, which i calculated from `SentryApp`, `SentryAppInstallation`, and `SentryAppInstallationToken`, which thus all need to generate outboxes on their own updates to keep the ApiToken's value fresh.  This results in some minor bits of refactoring to support.

### CODE REVIEW

File: src/sentry/api/endpoints/api_authorizations.py [hunk 1]
```diff
@@ -47,10 +48,11 @@ def delete(self, request: Request) -> Response:
         except ApiAuthorization.DoesNotExist:
             return Response(status=404)
 
-        with transaction.atomic(using=router.db_for_write(ApiToken)):
-            ApiToken.objects.filter(
+        with outbox_context(transaction.atomic(using=router.db_for_write(ApiToken)), flush=False):
```
Comment (corps) [inline src/sentry/api/endpoints/api_authorizations.py:51]: Now that ApiTokens are replicated, we must ensure that these operations generate outboxes correctly.

File: src/sentry/api/endpoints/api_tokens.py [hunk 2]
```diff
@@ -82,7 +84,11 @@ def delete(self, request: Request):
         if not token:
             return Response({"token": ""}, status=400)
 
-        ApiToken.objects.filter(user_id=user_id, token=token, application__isnull=True).delete()
+        with outbox_context(transaction.atomic(router.db_for_write(ApiToken)), flush=False):
```
Comment (corps) [inline src/sentry/api/endpoints/api_tokens.py:87]: Same same, handling outbox processing for the model.

File: src/sentry/db/models/outboxes.py [hunk 3]
```diff
@@ -365,6 +365,8 @@ def outbox_region_names(self) -> Collection[str]:
         """
         if hasattr(self, "organization_id"):
             return find_regions_for_orgs([self.organization_id])
+        if hasattr(self, "user_id"):
```
Comment (corps) [inline src/sentry/db/models/outboxes.py:368]: Makes it a bit easier to infer the regions for a replicated model that have a user id associated.

File: src/sentry/db/models/paranoia.py [hunk 4]
```diff
@@ -24,7 +24,7 @@ def get_queryset(self) -> ParanoidQuerySet:
         return ParanoidQuerySet(self.model, using=self._db).filter(date_deleted__isnull=True)
 
 
-class ParanoidModel(Model):
+class ParanoidModel(BaseModel):
```
Comment (corps) [inline src/sentry/db/models/paranoia.py:27]: To share `ParanoidModel` with `ReplicatedModel`, only one of them can subclass from `Model` or you get field conflicts in the meta class.  Can verify that this is a non impact in practice.

File: src/sentry/deletions/defaults/apigrant.py [hunk 5]
```diff
@@ -11,3 +16,7 @@ class ModelApiGrantDeletionTask(ModelDeletionTask):
     def mark_deletion_in_progress(self, instance_list):
         # no status to track
         pass
+
+    def delete_instance(self, instance):
+        with unguarded_write(router.db_for_write(ApiGrant)):
```
Comment (corps) [inline src/sentry/deletions/defaults/apigrant.py:21]: `ApiGrant` has a cascade into `SentryAppInstallation` via `SentryApp`.  It's ok not to generate the outbox in this case -- the cascading will result in ApiToken itself being deleted which will reap by other means.

File: src/sentry/models/outbox.py [hunk 6]
```diff
@@ -285,7 +293,7 @@ class OutboxScope(IntEnum):
             OutboxCategory.PROJECT_UPDATE,
             OutboxCategory.ORGANIZATION_INTEGRATION_UPDATE,
             OutboxCategory.SEND_SIGNAL,
-            OutboxCategory.ORGAUTHTOKEN_UPDATE,
+            OutboxCategory.ORGAUTHTOKEN_UPDATE_USED,
```
Comment (corps) [inline src/sentry/models/outbox.py:296]: This existing category is only used to handle the `used` column.

File: src/sentry/models/outbox.py [hunk 7]
```diff
@@ -249,13 +252,18 @@ def infer_identifiers(
                     shard_identifier = model.id
                 elif hasattr(model, "organization_id"):
                     shard_identifier = model.organization_id
-                elif hasattr(model, "auth_provider_id"):
-                    shard_identifier = model.auth_provider_id
+                elif hasattr(model, "auth_provider"):
+                    shard_identifier = model.auth_provider.organization_id
```
Comment (markstory) [inline src/sentry/models/outbox.py:256]: Is `auth_provider_id` entirely unused now? I don't see it leaving anywhere else in this diff.

Comment (corps) [reply src/sentry/models/outbox.py:256]: No, but the shard identifier should be the related organization_id and not the auth_provider_id itself.  This was a minor bug I noticed along the way.  It doesn't have much impact currently in practice.

File: src/sentry/db/models/paranoia.py (hunk 4 above)
Comment (markstory) [reply src/sentry/db/models/paranoia.py:27]: Could this switch in base classes impact backup/import as all the hook methods that backup uses are defined there.

Comment (markstory) [reply src/sentry/db/models/paranoia.py:27]: Read this again, and I get it now.

**Review by markstory — APPROVED** [review]

**Pull request merged**